hepstats/hypotests/calculators/frequentist_calculator.py

```diff
@@ -0,0 +1,430 @@
+        """
+        Create sampler with models.
+
+        Args:
+            floatting_params (list): floating parameter in the sampler
+
+        Example with `zfit`:
+            >>> sampler = calc.sampler(floatting_params=[zfit.Parameter("mean")])
+        """
+        self.set_dependents_to_bestfit()
+        nevents = []
+        for m, d in zip(self.model, self.data):
+            if m.is_extended:
+                nevents.append("extended")
+            else:
+                nevents.append(get_nevents(d))
+
+        return self._sampler(self.model,  nevents, floatting_params, *args, **kwargs)
+
+    def sample(self, sampler, ntoys, parameter=None, value=None):
+        """
+        Returns the samples generated from the sampler for a given value of a parameter
+
+        Args:
+            sampler (list): generator of samples
+            ntoys (int): number of samples to generate
+            parameter (optionnal): floating parameter in the sampler
+            value (optionnal): value of the parameter
+
+        Example with `zfit`:
+            >>> mean = zfit.Parameter("mean")
+            >>> sampler = calc.sampler(floatting_params=[mean])
+            >>> sample = calc.sample(sampler, 1000, mean, 1.2)
+        """
+        return self._sample(sampler, ntoys, parameter, value)
+
+    def toys_loss(self, parameter):
+        """
+        Construct a loss function constructed with a sampler for a given floating parameter
+
+        Args:
+            parameter: floating parameter in the sampler
+        Returns:
+             Loss function
+
+        Example with `zfit`:
+            >>> loss = calc.toys_loss(zfit.Parameter("mean"))
+        """
+        if parameter.name not in self._toys_loss:
+            sampler = self.sampler(floatting_params=[parameter])
+            self._toys_loss[parameter.name] = self.lossbuilder(self.model, sampler)
+        return self._toys_loss[parameter.name]
+
+    def _generate_and_fit_toys(self, poigen, ntoys, poieval, printfreq=0.2):
+        """
+        Generate and fit toys for a given POI.
+
+        Args:
+            poigen (POI): POI used to generate the toys
+            ntoys (int): number of toys to generate
+            poieval (POIarray): POI values to evaluate the loss function
+            printfreq: print frequency of the toys generation
+        """
+
+        minimizer = self.minimizer
+
+        param = poigen.parameter
+        gen_value = poigen.value
+
+        toys_loss = self.toys_loss(param)
+        sampler = toys_loss.data
+
+        result = {"bestfit": {"values": np.empty(ntoys), "nll": np.empty(ntoys)},
+                  "nll": {p: np.empty(ntoys) for p in poieval}}
+
+        printfreq = ntoys * printfreq
+
+        toys = self.sample(sampler, int(ntoys*1.2), param, gen_value)
+
+        for i in range(ntoys):
+            converged = False
+            toprint = i % printfreq == 0
+            while converged is False:
```

Comment:
This loop will never run because converged is False from the onset ;-).

Reply:
The `while` loop? Yes it does run.

Reply:
OK. But never the first time round, meaning for `i=0`. I guess that's intended, then. Thanks.

Reply:
Well at `i=0` , `converged = False` so it can enters the `while` loop. The purpose of this `while` loop is to keep only good toys, where the fit has converged. In principle I could ask to try to refit another time the toy to see if the minimum was not found in the first place.

Reply:
Oops, my brain tripped - of course you're right. Sorry!